File: content/examples/other/leetcode/problems/set-matrix-zeroes/example-2/solution.js
function setZeroes(matrix) {
  const rows = matrix.length;
  const cols = rows === 0 ? 0 : matrix[0].length;
  let firstRowZero = false;
  let firstColZero = false;

  for (let c = 0; c < cols; c += 1) if (matrix[0][c] === 0) firstRowZero = true;
  for (let r = 0; r < rows; r += 1) if (matrix[r][0] === 0) firstColZero = true;

  for (let r = 1; r < rows; r += 1) {
    for (let c = 1; c < cols; c += 1) {
      if (matrix[r][c] === 0) {
        matrix[r][0] = 0;
        matrix[0][c] = 0;
      }
    }
  }

  for (let r = 1; r < rows; r += 1) {
    for (let c = 1; c < cols; c += 1) {
      if (matrix[r][0] === 0 || matrix[0][c] === 0) matrix[r][c] = 0;
    }
  }

  if (firstRowZero) for (let c = 0; c < cols; c += 1) matrix[0][c] = 0;
  if (firstColZero) for (let r = 0; r < rows; r += 1) matrix[r][0] = 0;
  return matrix;
}

if (require.main === module) {
  console.log(setZeroes([[1, 1, 1], [1, 0, 1], [1, 1, 1]]));
}

module.exports = { setZeroes };

File: content/examples/other/leetcode/problems/set-matrix-zeroes/example-2/solution.py
def set_zeroes(matrix):
    rows = len(matrix)
    cols = len(matrix[0]) if rows else 0
    first_row_zero = any(matrix[0][c] == 0 for c in range(cols)) if rows else False
    first_col_zero = any(matrix[r][0] == 0 for r in range(rows)) if cols else False

    for r in range(1, rows):
        for c in range(1, cols):
            if matrix[r][c] == 0:
                matrix[r][0] = 0
                matrix[0][c] = 0

    for r in range(1, rows):
        for c in range(1, cols):
            if matrix[r][0] == 0 or matrix[0][c] == 0:
                matrix[r][c] = 0

    if first_row_zero and rows:
        for c in range(cols):
            matrix[0][c] = 0
    if first_col_zero and cols:
        for r in range(rows):
            matrix[r][0] = 0
    return matrix


if __name__ == "__main__":
    print(set_zeroes([[1, 1, 1], [1, 0, 1], [1, 1, 1]]))
